### PR TITLE
[AIRFLOW-5117] support refreshing EKS api tokens

### PR DESCRIPTION
**Pull request by houqp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5117) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5117

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR patches upstream k8s client to check for EKS api token refresh. Without this, Airflow K8S integration won't work if not deployed in cluster.

When Airflow scheduler or worker is not deployed in K8S cluster, all K8S API call will fail with 401 error after running for about 14min.

This is due to EKS enforces a 15min timeout for all API tokens but the official k8S python client doesn't support token refresh.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

see: tests/kubernetes/test_client.py

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by houqp** [comment]
upstream patch for swagger-codegen: https://github.com/swagger-api/swagger-codegen/pull/9624. I will work the PR for k8s client once the swagger-codegen PR is approved and merged.

**Comment by houqp** [comment]
@ashb tests added, ready for review :)

**Comment by houqp** [comment]
Looks like upstream k8s client is moving to a new openapi code generator, I have sent my code gen fix to openapi-generator as well. It will probably take awhile for the fix to land in upstream k8s client given they are still in the process of migrating to openapi-generator.

All my upstream fix activities can be tracked at https://github.com/kubernetes-client/python/issues/741.

**Comment by dimberman** [comment]
@houqp I'm a bit conflicted about this. On the one hand this is a very important bug that we need to address, on the other hand I'd want to minimize how much k8s logic is in airflow. Would it be possible to further isolate this so it would be easier to remove once the necessary fixes are implemented?

**Comment by houqp** [comment]
@dimberman my original patch added comment for every function and Class that can be safely removed once patch lands in upstream. I have pushed a new change to move most of the patching logic into a separate module to make it easier to follow.

File: scripts/ci/pylint_todo.txt [hunk 1]
```diff
@@ -201,6 +201,7 @@
 ./airflow/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
 ./airflow/kubernetes/kubernetes_request_factory/pod_request_factory.py
 ./airflow/kubernetes/kube_client.py
+./airflow/kubernetes/refresh_config.py
```
Comment (ashb) [inline scripts/ci/pylint_todo.txt]: Please don't add new files in to pylint todo, instead fix whatever pylint is complaining about :)

Comment (dimberman) [reply scripts/ci/pylint_todo.txt]: +1

File: airflow/kubernetes/refresh_config.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+class RefreshKubeConfigLoader(KubeConfigLoader):
+    """
+    Patched KubeConfigLoader, this subclass takes expirationTimestamp into
+    account and sets api key refresh callback hook in Configuration object
+    """
+    def __init__(self, *args, **kwargs):
+        KubeConfigLoader.__init__(self, *args, **kwargs)
+        self.api_key_expire_ts = None
+
+    def _load_from_exec_plugin(self):
```
Comment (dimberman) [inline airflow/kubernetes/refresh_config.py:44]: please write a comment explaining this function

**Review by dimberman — DISMISSED** [review]
Other than @ashb 's pylint comment and a a nit on my end this LGTM

File: scripts/ci/pylint_todo.txt (hunk 1 above)
Comment (houqp) [reply scripts/ci/pylint_todo.txt]: I initially tried fixing all the pylint errors, but the code turned to be much different than the original version. So I added it here to keep it similar to original k8s code to it's easier to compare with upstream code for further changes if needed. Turns out k8s python client doesn't pass our pylint rule ;)

With that, do you still want me to fix all the pylint error? If so, I can throw in my original patches.

Comment (ashb) [reply scripts/ci/pylint_todo.txt]: What is the pylint error in this new file? It should be possible to fix this file without changing anything else.

Comment (houqp) [reply scripts/ci/pylint_todo.txt]: I found a better way to handle this, with the new approach, we won't have to add new entry to pylint_todo.txt and we get to preserve the source code structure of upstream k8s client for easier diffing going forward.

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
@houqp Could you rebase please?

File: airflow/kubernetes/kube_client.py [hunk 3]
```diff
@@ -15,12 +15,48 @@
...
+
+from typing import Optional
 from airflow.configuration import conf
 
 try:
     from kubernetes import config, client
     from kubernetes.client.rest import ApiException  # pylint: disable=unused-import
+    from kubernetes.client.api_client import ApiClient
+    from kubernetes.client import Configuration
+    from airflow.kubernetes.refresh_config import (  # pylint: disable=ungrouped-imports
```
Comment (ashb) [inline airflow/kubernetes/kube_client.py:28]: `isort`, which we have now added might try to re-order this, so you'll probably have to add `# isort:skip` here too.

**Review by ashb — CHANGES_REQUESTED** [review]
Sorry, one further change. If you could rebase on to latest master, and then ping me when tests are green?

**Comment by houqp** [comment]
@ashb rebased and passed CI.

**Review by ashb — APPROVED** [review]

**Comment by danielpoonwj** [comment]
hi thank you for this fix!

I've upgraded to 1.10.7 specifically for this fix but I realized that the datetime format for my expirationTimestamp is `%Y-%m-%dT%H:%M:%SZ` (eg. `2020-01-05T10:00:00Z`) instead of `%Y-%m-%dT%H:%M:%S%z`.

it fails to parse and swallows the error with logging and essentially reverts to the previous behaviour.

was just wondering if its a percularity in my setup where others are actually receiving the timestamp in that format, or if it's a bug.

I can submit a fix depending if we need to support both datetime formats, or if we need to change it.

for reference, I am running airflow out of cluster on EC2 (Ubuntu 16.04) interacting with kubernetes v1.14.9 on EKS. awscli v1.16.310

**Comment by bhtucker** [comment]
Hey @danielpoonwj  -- 

This appears to be a Python 3.7 dependency. The problem is the same as in [this SO question](https://stackoverflow.com/questions/53291250/python-3-6-datetime-strptime-returns-error-while-python-3-7-works-well) .

Looks like there's already an issue for this: https://issues.apache.org/jira/browse/AIRFLOW-6585 and a PR: https://github.com/apache/airflow/pull/7153

**Pull request merged**